1.6.4 Protocol

brenoxavier opened the pull request:
This is my attempt to continue the work started by this PR #443.

I've written the specifications for protocol 78 (pre Netty) in version 1.6.4, but I have some doubts about whether this is correct because some packages are quite different from the post-Netty versions.

By setting the strings to UTF-16, I found that the protodef-validator is failing to validate the latest version of ProtoDef.

Used documentation:

[Protocol](https://wiki.vg/index.php?title=Protocol&oldid=4899)
[Data Types](https://wiki.vg/index.php?title=Data_types&oldid=4806)
[Metadata](https://wiki.vg/index.php?title=Entity_metadata&oldid=4900)
[Slot Data](https://wiki.vg/index.php?title=Slot_Data&oldid=4753)
[Object Data](https://wiki.vg/index.php?title=Object_Data&oldid=4754)

wgaylord commented:
It would be best to fix the tests so that they pass.

wgaylord commented:
Looks like an issue here https://github.com/ProtoDef-io/ProtoDef/blob/master/schemas/utils.json Where pstring only takes the count type and not encoding in the schema.  I have opened an issue on the ProtoDef repo for it. https://github.com/ProtoDef-io/ProtoDef/issues/51

wgaylord commented:
Looks like https://github.com/ProtoDef-io/node-protodef-validator/pull/11 will fix this.

extremeheat commented:
Can you be specific here with 1.6.4 instead of 1.6? Or were there no changes in protocol between then?

brenoxavier commented:
> Can you be specific here with 1.6.4 instead of 1.6? Or were there no changes in protocol between then?

I couldn't identify major differences in the protocol. I believe the protocol from 1.6.4 should work for all versions of 1.6.

https://wiki.vg/index.php?title=Protocol&type=revision&diff=4657&oldid=1096

extremeheat commented:
There should be any ambiguity there, if there is a protocol version change then there likely is a difference.

extremeheat commented on data/pc/1.6/version.json:
This needs to be added to the protocolVersions.json file

brenoxavier replied:
The version is already present in the file, does it need any other adjustments?

![image](https://github.com/PrismarineJS/minecraft-data/assets/41960989/67e7007f-6da9-476c-b2a9-76cd47648576)

brenoxavier commented:
> There should not be any ambiguity there, if there is a protocol version change then there likely is a difference.

Ok, I'll specify it as 1.6.4.

extremeheat commented:
What version did you use as a base to writing this?

brenoxavier commented:
> What version did you use as a base to writing this?

I used 1.7

extremeheat commented:
Yeah, I'm diffing 1.6.4 protocol.json to 1.7 and it seems there are alot of differences. Did the protocol really change that much or did you re-organize things? That would make review much harder as would need to go through all the packets individually

wgaylord commented:
> Yeah, I'm diffing 1.6.4 protocol.json to 1.7 and it seems there are alot of differences. Did the protocol really change that much or did you re-organize things? That would make review much harder as would need to go through all the packets individually

Yes, because 1.6 4 is pre-netty.

brenoxavier commented:
> Yeah, I'm diffing 1.6.4 protocol.json to 1.7 and it seems there are alot of differences. Did the protocol really change that much or did you re-organize things? That would make review much harder as would need to go through all the packets individually

Many things changed from 1.6 to 1.7, I tried to keep them as close as possible.

extremeheat commented:
I see. So yes this would require a more thorough review. Since there is no concept of states or clientbound/serverbound packets, it would probably be cleaner to not use the 1.7 protocol as reference (as its structure assumes those things, like toServer/toClient/play/etc).

1.7.6-1.7.10 protocol - https://wiki.vg/index.php?title=Protocol&oldid=6003
1.6.4 protocol - https://wiki.vg/index.php?title=Protocol&oldid=4899

brenoxavier commented:
So is it better to specify packages in just a single structure?

brenoxavier commented:
> I see. So yes this would require a more thorough review. Since there is no concept of states or clientbound/serverbound packets, it would probably be cleaner to not use the 1.7 protocol as reference (as its structure assumes those things, like toServer/toClient/play/etc).
> 
> 1.7.6-1.7.10 protocol - https://wiki.vg/index.php?title=Protocol&oldid=6003
> 1.6.4 protocol - https://wiki.vg/index.php?title=Protocol&oldid=4899

So is it better to specify packages in just a single structure?

extremeheat commented:
The point of that separation is so node-Minecraft-protocol can swap serializers when the state changes. That's not necessary for 1.6, so there should be one mapper for switching between packets (in the single state).

brenoxavier commented:
> The point of that separation is so node-Minecraft-protocol can swap serializers when the state changes. That's not necessary for 1.6, so there should be one mapper for switching between packets (in the single state). It technically could be separate in the data here but it doesn't make sense because the implementation in nmp would have to dedupe the switches.

Would a structure like this be a good idea? Segregate only into client packages, server packages, and packages used by both.

`{
  "types": {

  },
  "toClient": {

  },
  "toServer": {

  },
  "twoWay": {

  }
}`

extremeheat commented:
The point of protocol.json is something that protodef can read and use inside node-minecraft-protocol. It's not designed around documenting anything for sake of it (like what direction the packets are). It just needs to work inside node-minecraft-protocol. You can make an accompanying PR to node-minecraft-protocol if you think you can make it work that way. What I'm saying is I don't see how that will work without special handling for merging the main packet switch, at least not in nice way in the code, as it would require some preprocessing to make it work your way.

brenoxavier commented:
> The point of protocol.json is something that protodef can read and use inside node-minecraft-protocol. It's not designed around documenting anything for sake of it (like what direction the packets are). It just needs to work inside node-minecraft-protocol. You can make an accompanying PR to node-minecraft-protocol if you think you can make it work that way. What I'm saying is I don't see how that will work without special handling for merging the main packet switch, at least not in nice way in the code, as it would require some preprocessing to make it work your way.
> 
> What I meant was you just need { types, packets: {types} } and in node-minecraft-protocol you would do .addTypes(protocol.types); .addTypes(protocol.packets.types) to utilize it. Since there would be no collisions there would be no problem that way. Or just put them all directly into the root types.

Perfect, I'll simplify the packet structure and work to make node-minecraft-protocol compatible with it.

wgaylord commented:
I see some large issues so far. Why are their varints anywhere in the protocol.json. 1.6.4 didn't use varints at all. 

Multiple packets don't line up even with Wiki.vg for 1.6.4 Example the packet_encryption_key_response packet you have the handling of the byte arrays all incorrect where you read the length but then tell the array that it need to read the length as a byte itself. (I think this might be from now knowing how to use ProtoDef to do that more then just errors.)  You also have this multiple times in the file?

packet_encryption_key_request - Doesn't match wiki.vg at all 

We may want to change some field names on packet_player_position_and_look due to the order changing between if the packet is going from server to client or client to server

packet_player_digging - Missing the status byte as the first argument

packet_bed - Missing a byte in it.

And those are only some of the issues I found at first glance, let alone the organizational issue of 1.6.4 not having server vs client or the concepts of states.

brenoxavier commented:
> I see some large issues so far. Why are their varints anywhere in the protocol.json. 1.6.4 didn't use varints at all. 
> 
> Multiple packets don't line up even with Wiki.vg for 1.6.4 Example the packet_encryption_key_response packet you have the handling of the byte arrays all incorrect where you read the length but then tell the array that it need to read the length as a byte itself. (I think this might be from now knowing how to use ProtoDef to do that more then just errors.)  You also have this multiple times in the file?
> 
> packet_encryption_key_request - Doesn't match wiki.vg at all 
> 
> We may want to change some field names on packet_player_position_and_look due to the order changing between if the packet is going from server to client or client to server
> 
> packet_player_digging - Missing the status byte as the first argument
> 
> packet_bed - Missing a byte in it.
> 
> And those are only some of the issues I found at first glance, let alone the organizational issue of 1.6.4 not having server vs client or the concepts of states. 
> 
> Also we may want to rename some packets to be consistent with the other versions, provided they do the same type of thing.

I will review the packages and fix this. This is my first contact with ProtoDef.

brenoxavier commented:
I made some changes to the protocol.json:

- Simplified the protocol specification by removing states and directions. (@extremeheat)
- Reviewed all packages and corrected those identified as incorrect (@wgaylord).
- Aligned the protocol to closely resemble version 1.7.10.

wgaylord commented:
> * Reviewed all packages and corrected those identified as incorrect (@wgaylord).
I will go thru all the packets sometime soon and compare to source. Hopefully everything will line up well.

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+            "type": [
+              "array",
+              {
+                "count": "recordCount",
```
This needs to be looked at, as in 1.6.4 these records are just positions relative to the explosion written as bytes. 

See https://wiki.vg/index.php?title=Protocol&oldid=4899#Explosion_.280x3C.29

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+            "type": "f32"
+          },
+          {
+            "name": "pitch",
```
This should be an unsigned byte `u8`

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+            "type": "bool"
+          },
+          {
+            "name": "entityId",
```
This is conditional on inventoryType equaling 11 according to the source other wise this field doesn't exist

brenoxavier replied:
Should I remove the field or add a conditional?

wgaylord replied:
It should be conditional using a switch I believe.

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+          }
+        ]
+      ],
+      "packet_item_data": [
```
I think this should be renamed since its only used for sending over the images for on Maps.

brenoxavier replied:
Something like "packet_map_data"?

wgaylord replied:
1.7+ uses just `packet_map` for the packet that sends the map items image, so that is probably the best name for it.

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+            "type": "i16"
+          },
+          {
+            "name": "text",
```
This is just raw data in the form of a byte array of the length its prefixed with.

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+            "type": "i8"
+          },
+          {
+            "name": "scoreName",
```
scoreName and value are are not read or written if action = 1 

See 1.6.4 source or 
https://wiki.vg/index.php?title=Protocol&oldid=4899#Update_Score_.280xCF.29 for more details.

brenoxavier replied:
How would the specification for this be?

wgaylord replied:
It would look something like this https://github.com/PrismarineJS/minecraft-data/tree/master/data/pc/1.7/protocol.json#L2042-L2068

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+          }
+        ]
+      ],
+      "packet_legacy_server_list_ping": [
```
This should just be packet_server_list_ping the server responds using a kick packet.

It has the following fields seen in source and at https://wiki.vg/Server_List_Ping#1.6

a byte that just always holds 1
a byte that for some reason holds the ID of the CustomPayloadPacket
a string - This is just a constanct in this case its always "MC|PingHost"
a short  Length of the rest of the data in the packet
a byte - protocol version
a string  - host name of the server
a int - port of the server

brenoxavier commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+                  },
+                  {
+                    "name": "nbtData",
+                    "type": "compressedNbt"
```
It's the same as version 1.7; I believe it should work.

wgaylord replied:
This does seem correct, just decompiled 1.7 to check and while it uses varint all over it still uses a short for the NBT Data length.

brenoxavier commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+              "container",
+              [
+                {
+                  "name": "intField",
```
Should it be renamed to something like "entityId"?

wgaylord replied:
maybe ownerId?  Since its not the ID of the entity but the player that placed / spawned it.

wgaylord commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2290 @@
+          }
+        ]
+      ],
+      "packet_server_list_ping": [
```
Still have the issue that this packet is not implemented at all here.

brenoxavier replied:
How should I proceed?

wgaylord replied:
I documented it here, https://github.com/PrismarineJS/minecraft-data/pull/840#discussion_r1473553320 although the best way would be to open up 1.6.4 source and do it following that.

brenoxavier commented on data/pc/1.6.4/protocol.json:
```diff
@@ -0,0 +1,2286 @@
+          "3": "f32",
+          "4": "string",
+          "5": "slot",
+          "6": [
```
I couldn't quite understand how this should be done; in the other versions, they are bitfields.

rom1504 commented:
hi @brenoxavier would you like to try and open a PR on nmp using this ? 
otherwise it's a bit hard to tell if this would work

brenoxavier commented:
> hi @brenoxavier would you like to try and open a PR on nmp using this ? otherwise it's a bit hard to tell if this would work

It won't work. I will analyze how to modify the nmp to make it work with this.

rom1504 commented:
ok this is old; not going to happen

feel free to ask to re open or open a new pr if you want
